cpp/dolfinx/fem/discreteoperators.h

```diff
@@ -123,4 +123,199 @@ void assemble_discrete_gradient(const fem::FunctionSpace& V0,
     mat_set(dofmap1->cell_dofs(c), dofmap0->cell_dofs(c), Ae);
   }
 }
+
+/// @brief Assemble an interpolation operator matrix
+///
+/// This function creates a discrete operator \f$A\f$ that interpolates
+/// a function \f$V_0\f$ into a compatible space \f$V_1\f$, i.e.
+/// \f$V_0 \rightarrow V_1\f$. If \f$u_0\f$ is the
+/// degree-of-freedom vector associated with \f$V_0\f$, then
+/// \f$u_1=Au_0\f$ where \f$u_1\f$ is the degrees-of-freedom vector for
+/// interpolating function in the \f$V_1\f$ space.
+///
+/// @warning This function relies on the user supplying appropriate
+/// input and output spaces. See parameter descriptions.
+///
+/// @param[in] V0 The space to interpolate from
+/// @param[in] V1 The space to interpolate to
+/// @param[in] mat_set A functor that sets values in a matrix
+template <typename T, typename U>
+void assemble_interpolation_matrix(const fem::FunctionSpace& V0,
+                                   const fem::FunctionSpace& V1, U&& mat_set)
+{
+  // Get mesh
+  auto mesh = V0.mesh();
+  assert(mesh);
+
+  // Mesh dims
+  const int tdim = mesh->topology().dim();
+  const int gdim = mesh->geometry().dim();
+
+  // Get elements
+  std::shared_ptr<const FiniteElement> element0 = V0.element();
+  assert(element0);
+  std::shared_ptr<const FiniteElement> element1 = V1.element();
+  assert(element1);
+
+  xtl::span<const std::uint32_t> cell_info;
+  if (element1->needs_dof_transformations()
+      or element0->needs_dof_transformations())
+  {
+    mesh->topology_mutable().create_entity_permutations();
+    cell_info = xtl::span(mesh->topology().get_cell_permutation_info());
+  }
+
+  // Get dofmaps
+  auto dofmap0 = V0.dofmap();
+  auto dofmap1 = V1.dofmap();
+
+  const xt::xtensor<double, 2> X = element1->interpolation_points();
+  // Get block sizes and dof transformation operators
+  const int bs0 = element0->block_size();
+  const int bs1 = element1->block_size();
+  const auto apply_dof_transformation0
+      = element0->get_dof_transformation_function<double>(false, false, false);
+  const auto apply_inverse_dof_transform1
+      = element1->get_dof_transformation_function<T>(true, true, false);
+
+  // Get sizes of elements
+  const std::size_t dim0 = element0->space_dimension() / bs0;
+  const std::size_t value_size_ref0 = element0->reference_value_size() / bs0;
+  const std::size_t value_size0 = element0->value_size() / bs0;
+  const std::size_t value_size1 = element1->value_size() / bs1;
+
+  // Get geometry data
+  const fem::CoordinateElement& cmap = mesh->geometry().cmap();
+  const graph::AdjacencyList<std::int32_t>& x_dofmap
+      = mesh->geometry().dofmap();
+  // FIXME: Add proper interface for num coordinate dofs
```

Comment:
Let's stop reproducing this everywhere.

Reply:
I meant stop reproducing the bad code, not stop reproducing the message that the code is bad.

Reply:
Do you have a suggestion as to how you would like to see this resolved? As we currently use a constant width adjacency list for the dof map:
```
  const std::size_t num_dofs_g = x_dofmap.num_links(0);
 ```  
this gives us the number of dofs per cell.
If we want to make the code executable for P-refinement, we would need to move the definition of num dofs g and coordinate_dofs inside the cell loop.

Reply:
> Do you have a suggestion as to how you would like to see this resolved? As we currently use a constant width adjacency list for the dof map:
> 
> ```
>   const std::size_t num_dofs_g = x_dofmap.num_links(0);
> ```
> 
> this gives us the number of dofs per cell. If we want to make the code executable for mixed-topology meshes we would need to move the definition of num dofs g and coordinate_dofs inside the cell loop.

The dimension should probablt come from the `CoordinateElement`. ` x_dofmap.num_links(0)` takes the size from the first entry, but that requires the data to be non-empty.